server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java

```diff
@@ -118,4 +307,28 @@ public void afterBulk(long executionId, BulkRequest request, Throwable failure)
             }
         };
     }
+
+    private BulkProcessor.Listener countingListener(AtomicInteger requestCount, AtomicInteger successCount, AtomicInteger failureCount,
+                                                    AtomicInteger docCount) {
+
+        return new BulkProcessor.Listener() {
+            @Override
+            public void beforeBulk(long executionId, BulkRequest request) {
+                requestCount.incrementAndGet();
+            }
+
+            @Override
+            public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
+                successCount.incrementAndGet();
+                docCount.addAndGet(request.requests().size());
+            }
+
+            @Override
+            public void afterBulk(long executionId, BulkRequest request, Throwable failure) {
+                if (failure != null) {
+                    failureCount.incrementAndGet();
```

Comment:
We'd probably want to pass in the exceptionRef (and make it a throwable, but I thought we stopped catching throwable :|) and do the useOrSuppress thing here so we a view of all the exceptions during the test.